zkevm-circuits/src/copy_circuit.rs

```diff
@@ -468,6 +478,29 @@ impl<F: Field> SubCircuitConfig<F> for CopyCircuitConfig<F> {
     }
 }
 
+/// Verify that is_last goes to 1 at some point.
+pub fn constrain_must_terminate<F: Field>(
+    cb: &mut BaseConstraintBuilder<F>,
+    meta: &mut VirtualCells<'_, F>,
+    q_enable: Column<Fixed>,
+    tag: &BinaryNumberConfig<CopyDataType, 3>,
+) {
+    // If an event has started (tag != Padding on reader and writer rows), require q_enable=1 at the
+    // next step. This prevents querying rows where constraints are disabled.
+    //
+    // The tag is then copied to the next step by "rows[0].tag == rows[2].tag". Eventually,
+    // q_enable=0. By that point the tag must have switched to Padding, which is only possible with
+    // is_last=1. This guarantees that all the final conditions are checked.
+    let is_event = tag.value(Rotation::cur())(meta) - tag.constant_expr::<F>(CopyDataType::Padding);
+    cb.condition(is_event, |cb| {
+        cb.require_equal(
+            "the next step is enabled",
+            meta.query_fixed(q_enable, Rotation(2)),
+            1.expr(),
```

Comment:
It reads like this: if we are processing an event, then we are not about to hit the end of the circuit.

At the end of the circuit, q_enable switches to 0, which is how we detect the end. The rows just before that end are forced to be padding (is_event=0). This forces the last event to be fully verified by then.

Reply:
Yes, if the number of rows with `q_enabled = true` is constant (which it is), then I think this is all good.

Reply:
Hi @naure, what you say makes sense to me now, but I have a nitpick for readability.
It looks like if we unroll this constraint, it will be 

`is_event * (q_enable - 1)`

Which is the same as `(q_enable - 1) * is_event`. The latter way fits the Selector/Fixed column first convention.

Can we rewrite it to something like this for readability?
```
 cb.condition(meta.query_fixed(q_enable, Rotation(2)) - 1, |cb| {
        cb.require_equal(
            "The final step must be padding",
            is_event,
            0.expr(),
```